### PR TITLE
Spc

### PR DESCRIPTION
**Pull request by AodhanSweeney**
Adding SPC

### CODE REVIEW

**Comment by lesserwhirls** [comment]
I had no idea Siphon was failing in so many places. It's hard for me to tell for sure, but it seems like it's not been passing for some time now. The good news is that your two classes are not causing any of the errors at this point, which is good. 

The next step will be to write some tests for SPC and NCH. To avoid hitting live servers (which can go down and may not be reliable for testing purposes), we utilize `vcrpy` under the hood. Here is an example:

https://github.com/Unidata/siphon/blob/master/siphon/tests/test_wyoming.py

(the `vcrpy` library gets pulled in by `from siphon.testing import get_recorder`).

Basically, the first time you run the test code, any web request gets "recorded" by `vcrpy` in a "cassette" that will be played back each time you run the test in the future. These "cassettes" are stored in the github repository just like the code, and are found [here](https://github.com/Unidata/siphon/tree/master/siphon/tests/fixtures).

It's not super complicated, but it's a little tricky the first time you go through the testing processing like this.

**Comment by zbruick** [comment]
I'll get my hands dirty with siphon and work on fixing the Travis errors.

**Comment by AodhanSweeney** [comment]
Sorry guys just seeing this now but I will make sure to get on the tests.
Thanks for the heads up

**Comment by zbruick** [comment]
Note that this PR will close #259 and #257.

**Comment by zbruick** [comment]
The remaining Travis issue isn't caused by this PR (see discussion on #269). So as far as issues to fix, I think you're good for right now. I think this just needs to go to everyone (@lesserwhirls @dopplershift) for final review now.

**Comment by AodhanSweeney** [comment]
I saw that it also wanted a copyright so i gave it that as well, I think this is as far as I can take it though... Thanks for all the help though man! @zbruick

File: siphon/simplewebservice/iastate.py [hunk 1]
```diff
@@ -173,6 +173,6 @@ def _get_data_raw(self, time, site_id, pressure=None):
                 message += 'for station {stid}'.format(stid=site_id)
             if pressure is not None:
                 message += 'for pressure {pres}'.format(pres=pressure)
-            message = message[:-1] + '.'
+            message = message + '.'
```
Comment (lesserwhirls) [inline siphon/simplewebservice/iastate.py]: @zbruick - should this be here, or is this something that accidentally made it in this PR during the merge?

File: siphon/simplewebservice/nhc.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            unformatted_forecast_url = 'http://ftp.nhc.noaa.gov/atcf/archive/{}/a{}.dat.gz'
+            urlf = unformatted_forecast_url.format(year, filename[1:])
+            unformatted_best_url = 'http://ftp.nhc.noaa.gov/atcf/archive/{}/b{}.dat.gz'
+            urlb = unformatted_best_url.format(year, filename[1:])
+
+        url_links = [urlf, urlb]
+        url_count = 0
+        for url in url_links:
+            # Checking if url is valid, if status_code is 200 then website is active
+            if requests.get(url).status_code == 200:
```
Comment (lesserwhirls) [inline siphon/simplewebservice/nhc.py]: There might be cases where you get another value of 2xx code that is actually ok. Rather than checking for 200 exactly, you can use `raise_for_status()` on the response, and if something went wrong (4xx, 5xx), requests will raise a `requests.exceptions.HTTPError` error. See https://3.python-requests.org/user/quickstart/#response-status-codes.

You could replace the whole if block with something like:

~~~python
try:
    response = requests.get(url)
    response .raise_for_status()
    # rest of code after old if statement
    ...
except requests.exceptions.HTTPError as http_error:
    # give helpful message on console
    print("url {} was not valid, select different storm.".format(url))
    # raise the caught error
    raise    
~~~

Comment (lesserwhirls) [reply siphon/simplewebservice/nhc.py]: Now I'm wondering...since there is only one direct use of the requests library in your PR, perhaps class `NHCD` should subclass `HTTPEndPoint`, much like `IAStateUpperAir` does. @dopplershift @zbruick? Or is that something we refactor later, given the summer timeline?

File: siphon/simplewebservice/spc.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+Center's Archive of data that goes back to the 1950s.
+Weather events that are available are
+hail, wind, and tornados.
+
+"""
+
+import pandas as pd
+
+
+class SpcData:
```
Comment (lesserwhirls) [inline siphon/simplewebservice/spc.py]: `SpcData` should also use `HTTPEndPoint`, in the way that `NDBC` does. In this case, we would use the methods from `HTTPEndPoint` to read the data from a URL, and pass that to pandas `pd.read_csv` by wrapping the `HTTPEndPoint` return in a `StringIO`.  Again,  @dopplershift @zbruick, is that something we refactor later, given the summer timeline?

File: siphon/simplewebservice/iastate.py (hunk 1 above)
Comment (AodhanSweeney) [reply siphon/simplewebservice/iastate.py]: This, surprisingly, is a purposeful edit. It's in response to Error #257 in siphon about truncation of the final character in a string. I tested a few different scenarios and weird as it seems, the error seemed to be that the final character was continuously cut off.

Comment (lesserwhirls) [reply siphon/simplewebservice/iastate.py]: Awesome, just wanted to be sure. Sometimes little things squeek through a rebase and conflict resolve session.

File: siphon/simplewebservice/nhc.py (hunk 2 above)
Comment (zbruick) [reply siphon/simplewebservice/nhc.py]: Yeah I think consistent methods for accessing data whenever possible across the package would be nice. Since I haven't played around with any of this, I'm not sure how much refactoring it would take, but given how `IAStateUpperAir` uses it, I don't think it would take that much time. @AodhanSweeney, chime in if this would take too long - not sure what's left on your to-do list for the summer. If too much, we can refactor down the road.

Comment (zbruick) [reply siphon/simplewebservice/nhc.py]: We would still need to check if the URL was valid though, right? `NDBC` still does this with `_check_if_url_valid`, so I don't believe that's handled by `HTTPEndPoint`?

Comment (lesserwhirls) [reply siphon/simplewebservice/nhc.py]: I can think of two things to do to address this.

1) replace all `r.status_code == 200` checks with `r.raise_for_status()` (when you have a response obtained from `HTTPEndPoint`)

2) extend `HTTPEndPoint` to have a method for checking if a url is reachable (`is_url_reachable(url)`?) which simply does a head request on the url followed by a `raise_for_status()` (would replace all `_check_if_url_valid` type things). Maybe something like:

~~~python
def is_url_reachable(self, path, params=None):
        """Test if a url is reachable.

        Make a HEAD request, optionally including a parameters, to a path.
        The path of the request is the full URL.
        Parameters
        ----------
        path : str
            The URL to request
        params : DataQuery, optional
            The query to pass when making the request
        Returns
        -------
        resp : requests.Response
            The server's response to the request
        Raises
        ------
        HTTPError
            If the server returns a 4xx or 5xx code

        """
        resp = self._session.head(path, params=params)
        resp.raise_for_status()

        return true
~~~

Since `get` requests follow redirect by default, we should probably make sure the `head` call includes the parameter `allow_redirects=True` (`head` does not follow redirects by default), since the url would ultimately be reachable by a `get` request.

How does that sound?

**Comment by AodhanSweeney** [comment]
This most recent pull request is good for approval, but I want to work something else out so no one needs to look at it quite yet

**Comment by AodhanSweeney** [comment]
This is now ready for review, once this is in I will be pushing respective GUIs into the python library.

**Comment by AodhanSweeney** [comment]
This should be ready for review, it addresses issues #228 and #259

**Comment by lesserwhirls** [comment]
The build isn't passing at the moment, but outside of that, what remains on this one?

**Comment by dopplershift** [comment]
A review of the API for the NHC stuff (signature of public functions) to make sure we're ok with supporting that interface long term.
